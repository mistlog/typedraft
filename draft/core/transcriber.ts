import {
    Statement,
    FunctionDeclaration,
    BlockStatement,
    Expression,
    TemplateLiteral,
} from "@babel/types";
import { NodePath } from "@babel/traverse";
import { ToString } from "../common/utility";
import { ModuleCode } from "../code-object/module";
import { ExportClassCode } from "../code-object/export-class";
import { MethodCode } from "../code-object/method";
import { LocalContext } from "../code-object/local-context";
import { InlineContext } from "../code-object/inline-context";

/**
 * # Transcriber
 * We always write code in one language and translate(compile) it to another, however, in both language, the code we write directly represents the execution flow.
 * When we write draft, we don't expect it executes as is, instead, it's designed to be transformed, not executed.
 * That's why we choose another word to differentiate it from compiler.
 */
export class Transcriber {
    m_Module: ModuleCode;

    /**
     * the 4 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;

    constructor(_module: string) {
        this.m_Module = new ModuleCode(_module);

        this.m_ClassMap = new Map<string, ExportClassCode>();
        this.m_MethodMap = new Map<string, Array<MethodCode>>();
        this.m_ContextMap = new Map<string, LocalContext>();
        this.m_InlineContextMap = new Map<Symbol, InlineContext>();
        this.m_DSLMap = new Map<string, IDSL>();
        this.m_Plugins = new Array<IPlugin>();
    }

    /**
     * A transcriber is just a container, it's a collection of plugins.
     * For example, to transform local context, we have `draft-plugin-local-context`, to add class method to class, we have `draft-plugin-class`, and `draft-plugin-dsl` for DSL extension support.
     */

    Transcribe() {
        this.m_Plugins.forEach(plugin => plugin.Transcribe());
        return ToString(this.m_Module.m_File, {
            comments: false,
        });
    }

    AddDSL(name: string, dsl: IDSL) {
        this.m_DSLMap.set(name, dsl);
    }

    GetDSL(name: string) {
        return this.m_DSLMap.get(name);
    }

    GetLocalContext(name: string) {
        return this.m_ContextMap.get(name);
    }

    TraverseLocalContext(callback: TraverseLocalContextCallback) {
        this.m_ContextMap.forEach((context, name) => callback(context, name));
    }

    TraverseInlineContext(callback: TraverseInlineContextCallback) {
        this.m_InlineContextMap.forEach(context => callback(context));
    }

    GetClass(name: string) {
        return this.m_ClassMap.get(name);
    }

    TraverseMethod(callback: TraverseMethodCallback) {
        this.m_MethodMap.forEach((methods, class_name) => callback(methods, class_name));
    }
}

export interface IDSL {
    Transcribe?(
        block: Array<Statement>,
        path?: NodePath<FunctionDeclaration> | NodePath<BlockStatement>
    ): Array<Statement>;

    InplaceTranscribe?(literal: NodePath<TemplateLiteral>, type: string): Expression;

    m_Merge?: boolean;
}

export interface IPlugin {
    Transcribe(): void;
}

export interface ITranscriber {
    Transcribe: () => string;

    AddDSL: (name: string, dsl: IDSL) => void;

    GetDSL: (name: string) => IDSL;
    GetClass: (name: string) => ExportClassCode;
    GetLocalContext: (name: string) => LocalContext;

    TraverseMethod: (callback: TraverseMethodCallback) => void;
    TraverseLocalContext: (callback: TraverseLocalContextCallback) => void;
    TraverseInlineContext: (callback: TraverseInlineContextCallback) => void;

    m_Module: ModuleCode;

    /**
     * the 3 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;
}

export type TraverseLocalContextCallback = (context: LocalContext, name: string) => void;
export type TraverseInlineContextCallback = (context: InlineContext) => void;
export type TraverseMethodCallback = (methods: Array<MethodCode>, class_name: string) => void;
