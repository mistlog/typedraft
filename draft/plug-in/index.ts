export * from "./draft-plugin-class";
export * from "./draft-plugin-dsl";
export * from "./draft-plugin-filter";
export * from "./draft-plugin-inplace-context";
export * from "./draft-plugin-local-context";
export * from "./draft-plugin-refresh";
