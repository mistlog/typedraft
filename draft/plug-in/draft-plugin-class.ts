import { ITranscriber } from "../core";

export class ClassPlugin {
    m_Transcriber: ITranscriber;

    constructor(transcriber: ITranscriber) {
        this.m_Transcriber = transcriber;
    }

    Transcribe() {
        this.m_Transcriber.TraverseMethod((methods, class_name) =>
            methods.forEach(method =>
                this.m_Transcriber.GetClass(class_name).AddMember(method.ToClassMethod())
            )
        );
    }
}
