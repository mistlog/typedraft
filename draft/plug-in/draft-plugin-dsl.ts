import { ITranscriber } from "../core";
import { InlineContext, LocalContext } from "../code-object";

export class DSLPlugin {
    m_Transcriber: ITranscriber;

    constructor(transcriber: ITranscriber) {
        this.m_Transcriber = transcriber;
    }

    Transcribe() {
        const ResolveDSL = (context: InlineContext | LocalContext) => {
            const dsl = this.m_Transcriber.GetDSL(context.GetDSLName());

            if (dsl) {
                context.Resolve(dsl);
            }
        };

        this.m_Transcriber.TraverseInlineContext(ResolveDSL);
        this.m_Transcriber.TraverseLocalContext(ResolveDSL);
    }
}
