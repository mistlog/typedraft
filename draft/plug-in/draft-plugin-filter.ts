import { ITranscriber } from "../core";

export class FilterPlugin {
    m_Transcriber: ITranscriber;

    constructor(transcriber: ITranscriber) {
        this.m_Transcriber = transcriber;
    }

    Transcribe() {
        this.m_Transcriber.TraverseLocalContext(context => context.m_Path.remove());
        this.m_Transcriber.TraverseMethod(methods => methods.forEach(each => each.m_Path.remove()));
    }
}
