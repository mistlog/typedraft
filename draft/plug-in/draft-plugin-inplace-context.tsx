import { ITranscriber } from "../core";
import { ToString } from "../common";
import { MatchDSL, when, __, use } from "draft-dsl-match";

export class InplaceContextPlugin {
    m_Transcriber: ITranscriber;

    constructor(transcriber: ITranscriber) {
        this.m_Transcriber = transcriber;
    }

    Transcribe(this: InplaceContextPlugin) {
        const transcriber = this.m_Transcriber;
        transcriber.m_Module.Traverse({
            TaggedTemplateExpression(path) {
                Λ("match")` ${path.node} 
                    ${{
                        tag: {
                            type: "CallExpression",
                            typeParameters: use("typeParams"),
                            arguments: use("args"),
                            callee: {
                                type: "Identifier",
                                name: when(name => name === "context" || name === "Λ"),
                            },
                        },
                    }} -> ${(_, { args, typeParams }) => {
                    const dsl = transcriber.m_DSLMap.get(args[0].value);
                    if (dsl) {
                        const type = typeParams ? ToString(typeParams.params[0]) : "";
                        path.replaceWith(dsl.InplaceTranscribe(path.get("quasi"), type));
                    }
                }}

                    ${__} -> ${() => {}}
                `;
            },
        });
    }
}
