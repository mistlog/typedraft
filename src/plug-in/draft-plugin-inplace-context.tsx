import { ITranscriber } from "../core/transcriber";
import { NodePath, Node } from "@babel/core";
import { StringLiteral, Identifier } from "@babel/types";

export class InplaceContextPlugin {
    m_Transcriber: ITranscriber;
}

<InplaceContextPlugin /> +
    function constructor(this: InplaceContextPlugin, transcriber: ITranscriber) {
        this.m_Transcriber = transcriber;
    };

<InplaceContextPlugin /> +
    function Transcribe(this: InplaceContextPlugin) {
        const transcriber = this.m_Transcriber;
        transcriber.m_Module.Traverse({
            TaggedTemplateExpression(path) {
                const tag = path.get("tag");
                if (!tag.isCallExpression()) {
                    return;
                }

                const callee = tag.get("callee") as NodePath<Identifier>;
                const callee_name = callee.node.name as string;
                if (callee_name !== "context" && callee_name !== "Î›") {
                    return;
                }

                const [arg] = (tag.get("arguments") as Array<NodePath<Node>>) as [
                    NodePath<StringLiteral>
                ];
                const dsl_name = arg.node.value;

                const dsl = transcriber.m_DSLMap.get(dsl_name);
                if (!dsl) {
                    return;
                }

                path.replaceWith(dsl.InplaceTranscribe(path.get("quasi")));
            },
        });
    };
