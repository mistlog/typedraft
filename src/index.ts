export * from "./core/transcriber";

export * from "./code-object/export-class";
export * from "./code-object/local-context";
export * from "./code-object/inline-context";
export * from "./code-object/method";
export * from "./code-object/module";

export * from "./plug-in/draft-plugin-class";
export * from "./plug-in/draft-plugin-dsl";
export * from "./plug-in/draft-plugin-filter";
export * from "./plug-in/draft-plugin-local-context";
export * from "./plug-in/draft-plugin-refresh";

export * from "./common/utility";

/**
 *
 */
import { Transcriber } from "./core/transcriber";
import { RefreshDraftPlugin } from "./plug-in/draft-plugin-refresh";
import { DSLPlugin } from "./plug-in/draft-plugin-dsl";
import { LocalContextPlugin } from "./plug-in/draft-plugin-local-context";
import { ClassPlugin } from "./plug-in/draft-plugin-class";
import { FilterPlugin } from "./plug-in/draft-plugin-filter";

export function MakeDefaultTranscriber(_module: string) {
    const transcriber = new Transcriber(_module);
    transcriber.m_Plugins = [
        new RefreshDraftPlugin(transcriber),
        new DSLPlugin(transcriber),
        new RefreshDraftPlugin(transcriber),
        new LocalContextPlugin(transcriber),
        new ClassPlugin(transcriber),
        new FilterPlugin(transcriber),
    ];
    return transcriber;
}
