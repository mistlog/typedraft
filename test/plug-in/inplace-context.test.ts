import { MakeDefaultTranscriber, IDSL } from "../../src";
import { InplaceContextPlugin } from "../../src/plug-in/draft-plugin-inplace-context";
import { NodePath } from "@babel/core";
import { TemplateLiteral, stringLiteral } from "@babel/types";

class Math implements IDSL {
    InplaceTranscribe(literal: NodePath<TemplateLiteral>) {
        return stringLiteral("3");
    }
}

test("use inplace context as expression", () => {
    //
    const code = `
        const temp = Λ<number>("math")\`
            1+2
        \`;
    `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.m_Plugins = [new InplaceContextPlugin(transcriber)];
    transcriber.m_DSLMap.set("math", new Math());
    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

test("use inplace context as expression statement", () => {
    //
    const code = `
        context<number>("math")\`
            1+2
        \`;
    `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.m_Plugins = [new InplaceContextPlugin(transcriber)];
    transcriber.m_DSLMap.set("math", new Math());
    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

describe("not inplace context", () => {
    test("not call expression", () => {
        //
        const code = `
            context<number>\`
                1+2
            \`;
        `;

        const transcriber = MakeDefaultTranscriber(code);
        transcriber.m_Plugins = [new InplaceContextPlugin(transcriber)];
        const result = transcriber.Transcribe();
        expect(result).toMatchSnapshot();
    });

    test("not context or Λ", () => {
        //
        const code = `
            whatever<number>("math")\`
                1+2
            \`;
        `;

        const transcriber = MakeDefaultTranscriber(code);
        transcriber.m_Plugins = [new InplaceContextPlugin(transcriber)];
        const result = transcriber.Transcribe();
        expect(result).toMatchSnapshot();
    });

    test("no dsl", () => {
        //
        const code = `
            const temp = Λ<number>("math")\`
                1+2
            \`;
        `;

        const transcriber = MakeDefaultTranscriber(code);
        transcriber.m_Plugins = [new InplaceContextPlugin(transcriber)];
        const result = transcriber.Transcribe();
        expect(result).toMatchSnapshot();
    });
});
