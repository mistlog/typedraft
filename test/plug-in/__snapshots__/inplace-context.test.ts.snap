// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`not inplace context no dsl 1`] = `
"const temp = Λ<number>(\\"math\\")\`
                1+2
            \`;"
`;

exports[`not inplace context not call expression 1`] = `
"context<number>\`
                1+2
            \`;"
`;

exports[`not inplace context not context or Λ 1`] = `
"whatever<number>(\\"math\\")\`
                1+2
            \`;"
`;

exports[`use inplace context as expression 1`] = `"const temp = \\"3number\\";"`;

exports[`use inplace context as expression statement 1`] = `"\\"3\\";"`;
